# How do we sequence antibiotics

Though some antiotics are sythesized, some other antibiotics occur naturally. Bacteria and fungi produce antibiotics to kill bacteria. An example is penicillin produced by the penicillum mould and discovered by Alexander Flemming.  
Tyrocidine B1 is an antibiotics produced from the bactierm *Bacilus brevis*.  

Tyrocidine B1 unlike other normal human proteins are not synthesized through the normal central dogma process (of transcription and translation). They are known as non-ribosomal polypeptides as they are not produced by the ribosome but by a giant protein known as **NRP Synthetase**

  > NRP synthetase is a giant multi-module protein that assembles a cyclic peptide in  steps, one amino acid at a time. Each of ten different modules (shown by different  colors) adds a single amino acid to the peptide, which in the figure is one of  many tyrocidines produced by Bacilus brevis. In a final step, the peptide is circularized.
  
  
## Mass Spectometry
Since we cannot infer anything about the synthesis of Tyrocidine B1 by following the normal central dogma process, but by mass spectometry.
A mass spectometer is a molecular scale that shatters molecules to pieces and measures the size of the resulting fragments.

![Table of size of amino acids](http://bioinformaticsalgorithms.com/images/Antibiotics/integer_mass_table.png)
Tyrocidine B1, which is represented by VKLFPWFNQY, has total mass 1322 Da (99 + 128 + 113 + 147 + 97 + 186 + 147 + 114 + 128 + 163 = 1322).   

The collection of all masses generated by the mass spectrometer is known as an **experimental spectrum**.

To be able to identify the peptide from the mass spectometry spectrum, we would need to consider the **consitencies** of the spectrums of peptides and the mass of the polypeptide. By consistency, we mean peptides whose spectrum matches the spectrum generated by the mass spectrometer. The polypeptide in question would be determined by:
1. Being consistent with the spectrum of the mass spectrometer
2. Having the same mass `Mass(peptide)` as `ParentMass(spectrum)` which is the maximum mass of the spectrum from the mass spectrometer.

A brute force solution is not feasible for solving this, but we would use once called a `branch and bound` technique. This is an approach where we start from an empty string and then begin extending by the number of masses of amino acids and removing inconsistent peptides at each step.

Here is the algorithm:
```
CyclopeptideSequencing(Spectrum)
    CandidatePeptides ← a set containing only the empty peptide
    FinalPeptides ← empty list of strings
    while CandidatePeptides is nonempty
        CandidatePeptides ← Expand(CandidatePeptides)
        for each peptide Peptide in CandidatePeptides
            if Mass(Peptide) = ParentMass(Spectrum)
                if Cyclospectrum(Peptide) = Spectrum and Peptide is not in ﻿FinalPeptides
                    append Peptide to FinalPeptides
                remove Peptide from CandidatePeptides
            else if Peptide is not consistent with Spectrum
                remove Peptide from CandidatePeptides
    return FinalPeptides
```